### ISSUE TITLE
Problems with ACDCs whose memory setting was overwritten

### ISSUE THREAD

**Issue by haozturk** [dmwm/WMCore#10741]
**Impact of the bug**
The component which determines/monitors the memory threshold of requests, e.g. whoever enforces the 2000MB maxPSS in this example 
```
PerformanceKill (Exit code: 50660)
Error in CMSSW step cmsRun1
Number of Cores: 1
Job has exceeded maxPSS: 2000 MB
Job has PSS: 2192 MB
```

**Describe the bug**
For the past couple of weeks, I've realized that some ACDCs whose memory setting was overwritten are having issues. I see two issues in particular:

1. **Inconsistency between globally defined `Memory` parameter and the `Memory` defined within tasks.**

This is explained in the following issue in detail: https://github.com/CMSCompOps/WmAgentScripts/issues/891

2. **Memory setting of the ACDC workflow is not respected by the Performance Monitor.**

For instance, the following request's 1st task was submitted w/ 2GB per core, but it was not enough:
```
PerformanceKill (Exit code: 50660)
Error in CMSSW step cmsRun1
Number of Cores: 1
Job has exceeded maxPSS: 2000 MB
Job has PSS: 2223 MB
```
Then ACDCs were submitted w/ 2.3GB and 3GB of memory, but again the jobs were failed since they exceeded the previous threshold of 2GB of memory - so, the new 2.3GB and 3GB thresholds were not respected.

Requests: https://dmytro.web.cern.ch/dmytro/cmsprodmon/workflows.php?prep_id=task_SUS-RunIIFall17FSPremix-00218 

The original workflow: https://cmsweb.cern.ch/reqmgr2/fetch?rid=pdmvserv_task_SUS-RunIIFall17FSPremix-00218__v1_T_210703_150317_8449
Error: https://cms-unified.web.cern.ch/cms-unified/report/pdmvserv_task_SUS-RunIIFall17FSPremix-00218__v1_T_210703_150317_8449

First ACDC w/ 2300MB memory: https://cmsweb.cern.ch/reqmgr2/fetch?rid=cmsunified_ACDC0_task_SUS-RunIIFall17FSPremix-00218__v1_T_210803_121213_7312
Error: https://cms-unified.web.cern.ch/cms-unified/report/cmsunified_ACDC0_task_SUS-RunIIFall17FSPremix-00218__v1_T_210803_121213_7312

2nd ACDC w/ 3GB memory: https://cmsweb.cern.ch/reqmgr2/fetch?rid=cmsunified_ACDC1_task_SUS-RunIIFall17FSPremix-00218__v1_T_210805_143615_5865
Error: https://cms-unified.web.cern.ch/cms-unified/report/cmsunified_ACDC1_task_SUS-RunIIFall17FSPremix-00218__v1_T_210805_143615_5865  

**How to reproduce it**
Resubmitting the ACDCs mentioned above with the same configuration should reproduce the issue.

**Expected behavior**
We (P&R) should be able to set a custom memory value for a specific task and the new threshold for the new jobs of that task should be adjusted accordingly. As I explained above, there are two problems at the moment. 

1. Having 2 different memory settings (global one and nested one) is the first reason for confusion. It should be clear for us how to set the memory of a given task. With a dictionary defined globally or an absolute number defined within tasks.
2. Once we agree on this interface and propagate this information to the jobs, this should be applied.  

UPDATE (Alan): assignment validation function would fail with such dict value. 

**Additional context and error message**
Alan - please let me know if this long issue should be separated into two. Since the two issues that I reported here could be related, I kept them in the same issue. If this is not correct, I can separate things.

**Comment by amaltaro** [comment]
While watching how the new agent is behaving in production, I happen to spot one of these ACDC workflows where there is a mismatch between the memory requested in the workflow and what is actually taken into consideration in the workflow and in the job runtime.

Problem is with this workflow:
https://cmsweb.cern.ch/reqmgr2/fetch?rid=cmsunified_ACDC0_task_HIG-RunIISummer15wmLHEGS-05061__v1_T_210824_104244_7666

which has a top level memory set to (from the json document):
```
  "Memory": 2000,
```
and the same for the inner dictionary parameter (inside the tasks).


Looking into WMStats, I see it's failing a few jobs with:
```
WMBS job id: 13898
Task: /cmsunified_ACDC0_task_HIG-RunIISummer15wmLHEGS-05061__v1_T_210824_104244_7666/HIG-RunIISummer15wmLHEGS-05061_0
Agent: vocms0280.cern.ch
Errors:
    PerformanceError
        PerformanceKill (Exit Code: 50660)
            Error in CMSSW step cmsRun1
            Number of Cores: 1
            Job has exceeded maxPSS: 1000 MB
            Job has PSS: 1069 MB
```

which matches the memory requirements persisted in the workflow workload object (the spec file!).
```
cmsunified_ACDC0_task_HIG-RunIISummer15wmLHEGS-05061__v1_T_210824_104244_7666.tasks.HIG-RunIISummer15wmLHEGS-05061_0.input.splitting.performance.memoryRequirement = 1000.0
```

Having said and provided all that, we can now look into the Resubmission spec implementation in:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/StdSpecs/Resubmission.py#L44

in plain english. If the original workflow (the workflow which I created an ACDC for) was a TaskChain workflow, then:
* `Memory | Multicore | TimePerEvent` parameters must be of the type dictionary (usual TaskName : value key/value pairs), such that we can properly apply the new settings to the correct task name.

If we want to avoid this issue in the future, we could add a validation to the Resubmission spec to make it fail ACDCs for TaskChain when one of those parameters are provided in a "plain" data type, like string/integer/float.

@haozturk please let me know your thoughts.

**Comment by haozturk** [comment]
@amaltaro  In the past, we were providing memory in a dict and then we started to get error messages from WMCore side stating that memory should be int, not a dict. Please find the error message in the following issue description: https://github.com/CMSCompOps/WmAgentScripts/issues/891 

I could not hear from you back then and changed the logic in unified side and started providing memory in integer and apparently it does not work, either. I think we need a clear specification. Please let me know what think about the error message above.

**Comment by amaltaro** [comment]
Thanks for this pointer, Hasan. Indeed I can reproduce this error in my VM.

The memory validation function makes sense, but I thought we had a special validation process for such Resubmission workflows, such that a dict data type would be supported in this case.

I'm checking what are the best changes to be applied here and whether we can have a workaround for these workflows in production.

**Comment by haozturk** [comment]
Thanks Alan - IMO the outer level memory should be removed for taskchains and resubmission of taskchains - it's just confusing. We already have memory setting defined within tasks.

**Comment by amaltaro** [comment]
I need to think what's the best way to deal with this. 

The idea behind these resubmission workflows is that they are equal to their original workflow, besides very few exceptions that have been discussed in the past and that were made configurable during workflow creation.

Scanning the ReqMgr2 logs, I see this dictionary being provided during request creation:
```
[24/Aug/2021:10:42:44]  Updating request "None" with these user-provided args: {'Task1': {'PrepID': 'HIG-RunIISummer15wmLHEGS-05061', 'ProcessingString': 'MCRUN2_71_V1', 'ScramArch': ['slc6_amd64_gcc481'], 'SizePerEvent': 1468.0548072727272, 'ConfigCacheID': '8a1cef35266663756b608bf5fb35720d', 'Memory': 2000, 'TimePerEvent': 1.58015, 'AcquisitionEra': 'RunIISummer15wmLHEGS', 'PrimaryDataset': 'NonCascade_ggHToA1A2_A1ToBB_A2ToTauTau_MA2-60-MA1-30_Filter_TuneCUETP8M1_13TeV_madgraph_pythia8', 'CMSSWVersion': 'CMSSW_7_1_46_patch1', 'SplittingAlgo': 'EventBased', 'LheInputFiles': False, 'TaskName': 'HIG-RunIISummer15wmLHEGS-05061_0', 'KeepOutput': False, 'Campaign': 'RunIISummer15wmLHEGS', 'GlobalTag': 'MCRUN2_71_V1::All', 'Seeding': 'AutomaticSeeding', 'FilterEfficiency': 0.014727, 'Multicore': 1, 'EventsPerLumi': 6790, 'EventsPerJob': 20370, 'RequestNumEvents': 125000}, 'OriginalRequestName': 'pdmvserv_task_HIG-RunIISummer15wmLHEGS-05061__v1_T_210222_125552_6441', 'Task3': {'KeepOutput': False, 'MCPileup': '/Neutrino_E-10_gun/RunIISpring15PrePremix-PUMoriond17_80X_mcRun2_asymptotic_2016_TrancheIV_v2-v2/GEN-SIM-DIGI-RAW', 'PrepID': 'HIG-RunIISummer16DR80Premix-06423', 'InputFromOutputModule': 'PREMIXRAWoutput', 'GlobalTag': '80X_mcRun2_asymptotic_2016_TrancheIV_v6', 'TimePerEvent': 17, 'InputTask': 'HIG-RunIISummer16DR80Premix-06423_0', 'FilterEfficiency': 1, 'SplittingAlgo': 'EventAwareLumiBased', 'ProcessingString': 'PUMoriond17_80X_mcRun2_asymptotic_2016_TrancheIV_v6', 'ScramArch': ['slc6_amd64_gcc530'], 'SizePerEvent': 1750, 'ConfigCacheID': '8a1cef35266663756b608bf5fb31c527', 'Memory': 2000, 'Multicore': 8, 'TaskName': 'HIG-RunIISummer16DR80Premix-06423_1', 'AcquisitionEra': 'RunIISummer16DR80Premix', 'PrimaryDataset': 'NonCascade_ggHToA1A2_A1ToBB_A2ToTauTau_MA2-60-MA1-30_Filter_TuneCUETP8M1_13TeV_madgraph_pythia8', 'Campaign': 'RunIISummer16DR80Premix', 'CMSSWVersion': 'CMSSW_8_0_31'}, 'ACDCServer': 'https://cmsweb.cern.ch/couchdb', 'Campaign': 'RunIISummer15wmLHEGS', 'TimePerEvent': 0.1985, 'RequestType': 'Resubmission', 'OpenRunningTimeout': 0, 'InitialTaskPath': '/pdmvserv_task_HIG-RunIISummer15wmLHEGS-05061__v1_T_210222_125552_6441/HIG-RunIISummer15wmLHEGS-05061_0', 'ACDCDatabase': 'acdcserver', 'RequestString': 'ACDC0_task_HIG-RunIISummer15wmLHEGS-05061__v1_T', 'Requestor': 'cmsunified', 'Task4': {'KeepOutput': True, 'PrepID': 'HIG-RunIISummer16MiniAODv3-05842', 'InputFromOutputModule': 'AODSIMoutput', 'GlobalTag': '94X_mcRun2_asymptotic_v3', 'TimePerEvent': 0.1985, 'InputTask': 'HIG-RunIISummer16DR80Premix-06423_1', 'FilterEfficiency': 1, 'SplittingAlgo': 'EventAwareLumiBased', 'ProcessingString': 'PUMoriond17_94X_mcRun2_asymptotic_v3', 'ScramArch': ['slc6_amd64_gcc630'], 'SizePerEvent': 32.1363, 'ConfigCacheID': '8a1cef35266663756b608bf5fb31eda6', 'Memory': 2000, 'Multicore': 8, 'TaskName': 'HIG-RunIISummer16MiniAODv3-05842_0', 'AcquisitionEra': 'RunIISummer16MiniAODv3', 'PrimaryDataset': 'NonCascade_ggHToA1A2_A1ToBB_A2ToTauTau_MA2-60-MA1-30_Filter_TuneCUETP8M1_13TeV_madgraph_pythia8', 'Campaign': 'RunIISummer16MiniAODv3', 'CMSSWVersion': 'CMSSW_9_4_9'}, 'SizePerEvent': 32.1363, 'Memory': 2000, 'Group': 'ppd', 'Task2': {'KeepOutput': False, 'MCPileup': '/Neutrino_E-10_gun/RunIISpring15PrePremix-PUMoriond17_80X_mcRun2_asymptotic_2016_TrancheIV_v2-v2/GEN-SIM-DIGI-RAW', 'PrepID': 'HIG-RunIISummer16DR80Premix-06423', 'InputFromOutputModule': 'RAWSIMoutput', 'GlobalTag': '80X_mcRun2_asymptotic_2016_TrancheIV_v6', 'TimePerEvent': 17, 'InputTask': 'HIG-RunIISummer15wmLHEGS-05061_0', 'FilterEfficiency': 1, 'SplittingAlgo': 'EventAwareLumiBased', 'ProcessingString': 'PUMoriond17_80X_mcRun2_asymptotic_2016_TrancheIV_v6', 'ScramArch': ['slc6_amd64_gcc530'], 'SizePerEvent': 1750, 'ConfigCacheID': '8a1cef35266663756b608bf5fb319ab9', 'Memory': 2000, 'Multicore': 8, 'TaskName': 'HIG-RunIISummer16DR80Premix-06423_0', 'AcquisitionEra': 'RunIISummer16DR80Premix', 'PrimaryDataset': 'NonCascade_ggHToA1A2_A1ToBB_A2ToTauTau_MA2-60-MA1-30_Filter_TuneCUETP8M1_13TeV_madgraph_pythia8', 'Campaign': 'RunIISummer16DR80Premix', 'CMSSWVersion': 'CMSSW_8_0_31'}, 'RequestPriority': 180006, 'PrepID': 'task_HIG-RunIISummer15wmLHEGS-05061', 'CMSSWVersion': 'CMSSW_9_4_9'}
```

which only works because we have not enforced much validation on the Resubmission spec (and because it's very tricky, since it inherits data from any other possible spec type).

In short, creation of Resubmission workflows should have the very minimum information, according to the required parameters for a Resubmission spec type, e.g.:
```
{
    "ACDCDatabase": "acdcserver",
    "ACDCServer": "https://alancc7-cloud1.cern.ch/couchdb",
    "InitialTaskPath": "/amaltaro_TC_PY3_ProdPsi_Agent152_Val_210820_152013_9126/GenSimFull/GenSimFullMergeFEVTDEBUGoutput",
    "OriginalRequestName": "amaltaro_TC_PY3_ProdPsi_Agent152_Val_210820_152013_9126",
    "RequestString": "ACDC1_TC_PY3_ProdPsi_Agent152_Val",
    "RequestType": "Resubmission"
  }
```

and only what is supported and needs to be updated - based on the original request - should be provided, e.g.:
```
{
    "ACDCDatabase": "acdcserver",
    "ACDCServer": "https://alancc7-cloud1.cern.ch/couchdb",
    "InitialTaskPath": "/amaltaro_TC_PY3_ProdPsi_Agent152_Val_210820_152013_9126/GenSimFull/GenSimFullMergeFEVTDEBUGoutput",
    "Memory": {
      "GenSimFull": 1950,
      "Digi_2021noPU": 2950.0,
      "Reco_2021noPU": 3550
    },
    "OriginalRequestName": "amaltaro_TC_PY3_ProdPsi_Agent152_Val_210820_152013_9126",
    "RequestPriority": 500000,
    "RequestString": "ACDC1_TC_PY3_ProdPsi_Agent152_Val",
    "RequestType": "Resubmission",
    "TimePerEvent": {
      "GenSimFull": 0.5,
      "Digi_2021noPU": 2.0,
      "Reco_2021noPU": 1.1
    }
  }
```

Of course, we still have to fix this validate function, such that updates like:
```
    "Memory": {
      "GenSimFull": 1950,
      "Digi_2021noPU": 2950.0,
      "Reco_2021noPU": 3550
    }
```

can be properly parsed and checked. Many years ago, we had discussions in supporting a format like
```
"Task1": {"Memory": 1950},
"Task2": {"Memory": 2950},
"Task3": {"Memory": 3550}
```
but this was never implemented, and I'm slightly inclined to the former option. To be discussed as well.

**Comment by haozturk** [comment]
I'm happy either way. `Multicore` parameter should also be considered in this effort, I think. We should also be able to change core configuration in ACDC workflows. 

For the format, I think both of them works - if you'd ask me, the latter option looks better. Please let me know if you decide on a format, so that I can update the Unified code accordingly.

**Comment by haozturk** [comment]
Hi @amaltaro The workflows getting affected by this issue are delayed a lot and I'm told that the completion of these requests are getting quite urgent due to some publication deadlines. If you think that the fix of this issue will take time, can you think of any workaround for this?

**Comment by amaltaro** [comment]
I started working on this yesterday, Hasan. I'm trying to get something implemented, tested and deployed in testbed by tomorrow.

I finally understood why it suddenly stopped working, and the reason is that Python2 accepts comparison between `dict` and `int`, while Python3 no longer does. @mapellidario you probably knew this already, but it might pop up again in your future challenges and projects ;)

**Comment by jordan-martins** [comment]
Hi @amaltaro @haozturk, 
thanks a lot for working on this. We would very much appreciate it if a fix could be provided asap. We have 15 wfs from SUS that need this fix to run the recoveries properly! They are delaying publication at this point (7 weeks trying to run in highprio).

Thanks a lot,
Jordan
